File: examples/adler32/c/adler_drv.c
/* cando-style function-level differential driver — C ORACLE side.
 * Protocol: `driver <func> [args...]` → print the result CANONICALLY to stdout,
 * exit 0 on success, nonzero on error/unsupported. The Rust driver
 * (rust/src/bin/adler_drv.rs) prints byte-identically. Keep it THIN. */
#include <stdio.h>
#include <string.h>
#include "adler32.h"

int main(int argc, char **argv) {
    if (argc < 2) {
        fprintf(stderr, "usage: %s <func> [args...]\n", argv[0]);
        return 2;
    }
    if (strcmp(argv[1], "adler32") == 0) {
        if (argc < 3) return 3;   /* bad vector: adler32 needs a data arg */
        const char *s = argv[2];
        printf("%08x\n", adler32((const uint8_t *)s, strlen(s)));
        return 0;
    }
    fprintf(stderr, "unsupported function: %s\n", argv[1]);
    return 4;
}

File: examples/adler32/rust/src/bin/adler_drv.rs
//! cando-style function-level differential driver — RUST side. Same protocol as
//! `../../c/adler_drv.c`: `driver <func> [args...]` → canonical stdout, exit
//! 0/nonzero. No `unwrap()` on harness-fed argv — return a nonzero exit so a bad
//! vector is a clean baseline failure, not a panic.
use std::process::exit;

fn main() {
    let args: Vec<String> = std::env::args().skip(1).collect();
    let Some(func) = args.first().map(String::as_str) else {
        eprintln!("usage: driver <func> [args...]");
        exit(2);
    };
    match func {
        "adler32" => match args.get(1) {
            Some(data) => println!("{:08x}", radler::adler32_bytes(data.as_bytes())),
            None => exit(3),
        },
        other => {
            eprintln!("unsupported function: {other}");
            exit(4);
        }
    }
}
